pkg/collector/corechecks/systemd/systemd.go

```diff
@@ -0,0 +1,448 @@
+}
+
+func (c *SystemdCheck) submitMetrics(sender aggregator.Sender, conn *dbus.Conn) error {
+	units, err := c.stats.ListUnits(conn)
+	if err != nil {
+		return fmt.Errorf("Error getting list of units: %v", err)
+	}
+
+	c.submitCountMetrics(sender, units)
+
+	loadedCount := 0
+	for _, unit := range units {
+		if unit.LoadState == unitLoadedState {
+			loadedCount++
+		}
+		if !c.isMonitored(unit.Name) {
+			continue
+		}
+		tags := []string{"unit:" + unit.Name}
+		sender.ServiceCheck(unitStateServiceCheck, getServiceCheckStatus(unit.ActiveState), "", tags, "")
+
+		c.submitBasicUnitMetrics(sender, conn, unit, tags)
+		c.submitPropertyMetricsAsGauge(sender, conn, unit, tags)
+	}
+
+	sender.Gauge("systemd.unit.loaded.count", float64(loadedCount), "", nil)
+	return nil
+}
+
+func (c *SystemdCheck) submitBasicUnitMetrics(sender aggregator.Sender, conn *dbus.Conn, unit dbus.UnitStatus, tags []string) {
+	unitProperties, err := c.stats.GetUnitTypeProperties(conn, unit.Name, dbusTypeMap[typeUnit])
+	if err != nil {
+		log.Errorf("Error getting unit unitProperties: %s", unit.Name)
+		return
+	}
+
+	activeEnterTimestamp, err := getPropertyUint64(unitProperties, "ActiveEnterTimestamp")
+	if err != nil {
+		log.Errorf("Error getting property ActiveEnterTimestamp: %v", err)
+		return
+	}
+	active := 0
+	if unit.ActiveState == unitActiveState {
+		active = 1
+	}
+	loaded := 0
+	if unit.LoadState == unitLoadedState {
+		loaded = 1
+	}
+	sender.Gauge("systemd.unit.active", float64(active), "", tags)
+	sender.Gauge("systemd.unit.loaded", float64(loaded), "", tags)
+	sender.Gauge("systemd.unit.uptime", float64(computeUptime(unit.ActiveState, activeEnterTimestamp, c.stats.TimeNanoNow())), "", tags)
+}
+
+func (c *SystemdCheck) submitCountMetrics(sender aggregator.Sender, units []dbus.UnitStatus) {
+	counts := map[string]int{}
+
+	for _, activeState := range unitActiveStates {
+		counts[activeState] = 0
+	}
+
+	for _, unit := range units {
+		counts[unit.ActiveState]++
+	}
+
+	for _, activeState := range unitActiveStates {
+		count := counts[activeState]
+		sender.Gauge("systemd.unit.count", float64(count), "", []string{"active_state:" + activeState})
+	}
+}
+
+func (c *SystemdCheck) submitPropertyMetricsAsGauge(sender aggregator.Sender, conn *dbus.Conn, unit dbus.UnitStatus, tags []string) {
+	for unitType := range metricConfigs {
+		if !strings.HasSuffix(unit.Name, "."+unitType) {
+			continue
+		}
+		serviceProperties, err := c.stats.GetUnitTypeProperties(conn, unit.Name, dbusTypeMap[unitType])
+		if err != nil {
+			log.Errorf("Error getting serviceProperties for service: %s", unit.Name)
+			return
+		}
+		for _, service := range metricConfigs[unitType] {
+			err := sendServicePropertyAsGauge(sender, serviceProperties, service, tags)
+			if err != nil {
+				msg := fmt.Sprintf("Cannot send property '%s' for unit '%s': %v", service.propertyName, unit.Name, err)
+				if service.optional {
+					log.Debugf(msg)
+				} else {
+					log.Errorf(msg)
```

Comment:
typically, logging an `Error`-level entry for every property that can't be sent may be too verbose: a `warning` or `info`  level log is probably enough.